include/deal.II/differentiation/ad/sacado_number_types.h

```diff
@@ -0,0 +1,804 @@
+    {};
+
+
+    /**
+     * Specialization of the NumberTraits struct for
+     * the (otherwise disabled) nested complex Sacado number type.
+     */
+    template<>
+    struct NumberTraits<std::complex< Sacado::Fad::DFad<Sacado::Fad::DFad<float> > >,NumberTypes::sacado_dfad_dfad>
+      : NumberTraits<typename ADNumberTraits<std::complex< Sacado::Fad::DFad<Sacado::Fad::DFad<float> > > >::scalar_type,NumberTypes::sacado_dfad_dfad>
+    {};
+
+
+    /**
+     * Specialization of the NumberTraits struct for
+     * the (otherwise disabled) nested Sacado number type.
+     */
+    template<>
+    struct NumberTraits<Sacado::Fad::DFad<Sacado::Fad::DFad<double> >,NumberTypes::sacado_dfad_dfad>
+      : NumberTraits<typename ADNumberTraits<Sacado::Fad::DFad<Sacado::Fad::DFad<double> > >::scalar_type,NumberTypes::sacado_dfad_dfad>
+    {};
+
+
+    /**
+     * Specialization of the NumberTraits struct for
+     * the (otherwise disabled) nested complex Sacado number type.
+     */
+    template<>
+    struct NumberTraits<std::complex< Sacado::Fad::DFad<Sacado::Fad::DFad<double> > >,NumberTypes::sacado_dfad_dfad>
+      : NumberTraits<typename ADNumberTraits<std::complex< Sacado::Fad::DFad<Sacado::Fad::DFad<double> > > >::scalar_type,NumberTypes::sacado_dfad_dfad>
+    {};
+
+
+    /* -------------- NumberTypes::sacado_rad_dfad -------------- */
+
+    /**
+     * Specialization of the general ADNumberTraits class that
+     * provides relevant information for auto-differentiable numbers.
+     * This specialization is for the case where @p ADNumberType is an
+     * once nested Sacado::Fad::DFad number, with the inner number templated on
+     * a floating point type.
+     */
+    template<typename ADNumberType>
+    struct ADNumberTraits<ADNumberType, typename std::enable_if<
+      std::is_same<ADNumberType, Sacado::Rad::ADvar<Sacado::Fad::DFad<typename ADNumberType::ADVari::scalar_type> > >::value
+      >::type>
+      : NumberTraits<typename ADNumberType::ADVari::scalar_type, NumberTypes::sacado_rad_dfad>
+    {};
+
+
+    /**
+     * Specialization of the NumberTraits struct for
+     * the (otherwise disabled) nested Sacado number type.
+     */
+    template<>
+    struct NumberTraits<Sacado::Rad::ADvar<Sacado::Fad::DFad<float> >,NumberTypes::sacado_rad_dfad>
+      : NumberTraits<typename ADNumberTraits<Sacado::Rad::ADvar<Sacado::Fad::DFad<float> > >::scalar_type,NumberTypes::sacado_rad_dfad>
+    {};
+
+
+    /**
+     * Specialization of the NumberTraits struct for
+     * the (otherwise disabled) nested Sacado number type.
+     */
+    template<>
+    struct NumberTraits<Sacado::Rad::ADvar<Sacado::Fad::DFad<double> >,NumberTypes::sacado_rad_dfad>
+      : NumberTraits<typename ADNumberTraits<Sacado::Rad::ADvar<Sacado::Fad::DFad<double> > >::scalar_type,NumberTypes::sacado_rad_dfad>
+    {};
+
+
+#ifdef DEAL_II_TRILINOS_CXX_SUPPORTS_SACADO_COMPLEX_RAD
```

Comment:
@bangerth I decided not to go for the `static_assert` option as we discussed yesterday. It breaks various tangentially related things and I don't have the time right now to find a solution for this. I've made a note of this point [in the AD project board](https://github.com/dealii/dealii/projects/7) and will return to it later, if thats OK with you.

Reply:
Yes, of course. At your discretion.